etcd-recovery/etcdrecovery.go

```diff
@@ -0,0 +1,538 @@
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	crclient "sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/log/zap"
+)
+
+const (
+	defaultEtcdClientTimeout = 5 * time.Second
+	defaultDialTimeout       = 3 * time.Second
+	healthyClusterWait       = 5 * time.Minute
+)
+
+type options struct {
+	etcdClientCertFile string
+	etcdClientKeyFile  string
+	etcdCAFile         string
+	namespace          string
+}
+
+func NewRecoveryCommand() *cobra.Command {
+	opts := options{
+		etcdClientCertFile: "/etc/etcd/tls/client/etcd-client.crt",
+		etcdClientKeyFile:  "/etc/etcd/tls/client/etcd-client.key",
+		etcdCAFile:         "/etc/etcd/tls/etcd-ca/ca.crt",
+	}
+
+	cmd := &cobra.Command{
+		Use:          "recover-etcd",
+		Short:        "Commands to report on etcd status and recover member that is not functional",
+		SilenceUsage: true,
+		Run: func(cmd *cobra.Command, args []string) {
+			cmd.Help()
+			os.Exit(1)
+		},
+	}
+
+	cmd.PersistentFlags().StringVar(&opts.etcdClientCertFile, "etcd-client-cert", opts.etcdClientCertFile, "etcd client cert file.")
+	cmd.PersistentFlags().StringVar(&opts.etcdClientKeyFile, "etcd-client-key", opts.etcdClientKeyFile, "etcd client cert key file.")
+	cmd.PersistentFlags().StringVar(&opts.etcdCAFile, "etcd-ca-cert", opts.etcdCAFile, "etcd trusted CA cert file.")
+	cmd.PersistentFlags().StringVar(&opts.namespace, "namespace", "", "namespace of etcd cluster")
+	_ = cmd.MarkFlagRequired("etcd-endpoints")
+	_ = cmd.MarkFlagRequired("namespace")
+
+	cmd.AddCommand(NewStatusCommand(&opts))
+	cmd.AddCommand(NewRunCommand(&opts))
+
+	return cmd
+}
+
+func NewRunCommand(opts *options) *cobra.Command {
+	cmd := &cobra.Command{
+		Use:          "run",
+		Short:        "Recover etcd cluster that has quorum",
+		SilenceUsage: true,
+		Run: func(cmd *cobra.Command, args []string) {
+			ctx, cancel := signal.NotifyContext(cmd.Context(), os.Interrupt, syscall.SIGINT)
+			defer cancel()
+			ctx = setupCmdContext(ctx)
+			err := runRecovery(ctx, *opts)
+			if err != nil {
+				ctrl.Log.Error(err, "Error occurred")
+				os.Exit(1)
+			}
+		},
+	}
+
+	return cmd
+}
+
+func setupCmdContext(cmdContext context.Context) context.Context {
+	ctrl.SetLogger(zap.New(zap.JSONEncoder(func(o *zapcore.EncoderConfig) {
+		o.EncodeTime = zapcore.RFC3339TimeEncoder
+	})))
+	ctx := ctrl.LoggerInto(cmdContext, ctrl.Log.WithName("etcd-recovery"))
+	return ctx
+}
+
+type etcdStatus struct {
+	failingPod    *string
+	missingMember *string
+	recoverable   bool
+}
+
+func (s *etcdStatus) isUnhealthy() bool {
+	return s.missingMember != nil || s.failingPod != nil
+}
+
+func (s *etcdStatus) isRecoverable() bool {
+	recoverable := s.recoverable
+	if s.failingPod != nil && s.missingMember != nil {
+		if *s.failingPod != *s.missingMember {
+			recoverable = false
+		}
+	}
+	return recoverable
+}
+
+func (s *etcdStatus) memberToRecover() string {
+	if !s.isRecoverable() {
+		return ""
+	}
+	if s.failingPod != nil {
```

Comment:
why this one takes precedence?

Reply:
we just need *a* name... at this point we know that if both failingPod and failingMember are set, they have to match (otherwise isRecoverable returned false).